planner/core/logical_plan_builder.go

```diff
@@ -2473,6 +2550,88 @@ func (b *PlanBuilder) buildDelete(delete *ast.DeleteStmt) (Plan, error) {
 	return del, nil
 }
 
+func (b *PlanBuilder) buildProjectionForWindow(p LogicalPlan, byItems []*ast.ByItem, aggMap map[*ast.AggregateFuncExpr]int) (LogicalPlan, []property.Item, error) {
+	b.optFlag |= flagEliminateProjection
+	exprs := make([]property.Item, 0, len(byItems))
+	proj := LogicalProjection{Exprs: make([]expression.Expression, 0, len(p.Schema().Columns)+len(byItems))}.Init(b.ctx)
+	schema := expression.NewSchema(make([]*expression.Column, 0, len(p.Schema().Columns)+len(byItems))...)
+	for _, col := range p.Schema().Columns {
+		proj.Exprs = append(proj.Exprs, col)
+		schema.Append(col)
+	}
+	transformer := &itemTransformer{}
+	for _, item := range byItems {
+		newExpr, _ := item.Expr.Accept(transformer)
+		item.Expr = newExpr.(ast.ExprNode)
+		it, np, err := b.rewrite(item.Expr, p, aggMap, true)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		p = np
+		if col, ok := it.(*expression.Column); ok {
+			exprs = append(exprs, property.Item{Col: col, Desc: item.Desc})
+			continue
+		}
+		proj.Exprs = append(proj.Exprs, it)
+		col := &expression.Column{
+			ColName:  model.NewCIStr(fmt.Sprintf("%d_proj_window_%d", p.ID(), schema.Len())),
+			UniqueID: b.ctx.GetSessionVars().AllocPlanColumnID(),
+			RetType:  it.GetType(),
+		}
+		schema.Append(col)
+		exprs = append(exprs, property.Item{Col: col, Desc: item.Desc})
+	}
+	proj.SetSchema(schema)
+	proj.SetChildren(p)
+	return proj, exprs, nil
+}
+
+func (b *PlanBuilder) buildWindowFunction(p LogicalPlan, expr *ast.WindowFuncExpr, aggMap map[*ast.AggregateFuncExpr]int) (*LogicalWindow, error) {
+	spec := expr.Spec
+	// Add sort for partition by and order by
+	var items []*ast.ByItem
+	if spec.PartitionBy != nil {
+		items = append(items, spec.PartitionBy.Items...)
+	}
+	if spec.OrderBy != nil {
+		items = append(items, spec.OrderBy.Items...)
+	}
+	var byItems []property.Item
+	if len(items) > 0 {
+		var err error
+		p, byItems, err = b.buildProjectionForWindow(p, items, aggMap)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	newArgList := make([]expression.Expression, 0, len(expr.Args))
+	for _, arg := range expr.Args {
+		newArg, np, err := b.rewrite(arg, p, aggMap, true)
```

Comment:
We should also add a `Projection` if any of the arguments is a `ScalarFunction`.

Reply:
I think it is unnecessary, it just adds an extra burden to store the result in the chunk, we can just calculate the result and use it.

Reply:
No, it's better to let a `Project` operator to do all the scalar function evaluation, let the `Window` operator focused on the window function. 

This helps us to:
1. simplify the code logic of the `Window` operator. We only need to consider `Column`s in the window functions. This is what we have done in the aggregate functions.
2. separate the performance improvement between `Window` and `Project` operators.